airsonic-main/src/main/java/org/airsonic/player/service/search/SearchServiceImpl.java

```diff
@@ -0,0 +1,298 @@
+import org.apache.lucene.store.FSDirectory;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.io.IOException;
+import java.util.*;
+import java.util.function.BiConsumer;
+import java.util.stream.Collectors;
+
+import static org.airsonic.player.service.search.IndexType.*;
+import static org.springframework.util.ObjectUtils.isEmpty;
+
+@Service
+public class SearchServiceImpl implements SearchService {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SearchServiceImpl.class);
+
+  @Autowired
+  private QueryFactory queryFactory;
+  @Autowired
+  private IndexManager indexManager;
+  @Autowired
+  private SearchServiceTermination termination;
+
+  // TODO Should be changed to SecureRandom?
+  private final Random random = new Random(System.currentTimeMillis());
+
+  public SearchServiceImpl() {
+    removeLocks();
+  }
+
+  @Override
+  public void startIndexing() {
+    indexManager.startIndexing();
+  }
+
+  @Override
+  public void index(MediaFile mediaFile) {
+    indexManager.index(mediaFile);
+  }
+
+  @Override
+  public void index(Artist artist, MusicFolder musicFolder) {
+    indexManager.index(artist, musicFolder);
+  }
+
+  @Override
+  public void index(Album album) {
+    indexManager.index(album);
+  }
+
+  @Override
+  public void stopIndexing() {
+    indexManager.stopIndexing();
+  }
+
+  @Override
+  public SearchResult search(SearchCriteria criteria, List<MusicFolder> musicFolders,
+      IndexType indexType) {
+
+    SearchResult result = new SearchResult();
+    int offset = criteria.getOffset();
+    int count = criteria.getCount();
+    result.setOffset(offset);
+
+    if (count <= 0)
+      return result;
+
+    IndexReader reader = null;
+    try {
+
+      reader = indexManager.createIndexReader(indexType);
+      Searcher searcher = new IndexSearcher(reader);
+      Query query = queryFactory.search(criteria, musicFolders, indexType);
+
+      TopDocs topDocs = searcher.search(query, null, offset + count);
+      result.setTotalHits(topDocs.totalHits);
+      int start = Math.min(offset, topDocs.totalHits);
+      int end = Math.min(start + count, topDocs.totalHits);
+
+      for (int i = start; i < end; i++) {
+        termination.addIfAnyMatch(result, indexType, searcher.doc(topDocs.scoreDocs[i].doc));
+      }
+
+    } catch (IOException | ParseException e) {
+      LOG.error("Failed to execute Lucene search.", e);
+    } finally {
+      FileUtil.closeQuietly(reader);
+    }
+    return result;
+  }
+
+  /**
+   * Common processing of random method.
+   * @param count Number of albums to return.
+   * @param searcher
+   * @param query
+   * @param id2ListCallBack Callback to get D from id and store it in List
+   * @return result
+   * @throws IOException
+   */
+  private final <D> List<D> createRandomFiles(
```

Comment:
I think this method needs to be renamed. How about createRandomDocsList?

Reply:
I revised that according to your advice.